docs/developer/style_guide.md

```diff
@@ -0,0 +1,579 @@
+but only use to move a maximum of one level of nesting where possible. Do NOT
+use `goto`.
+
+
+### Functions
+
+A function subprogram should be "pure", i.e. generate no side-effects through
+modification of variables in argument list or module variables through "use
+association". Avoid recursion.
+
+
+### Use of `use`
+
+Use modules with an `only` clause if only a small number of variables
+required. Put appropriate `use` statements into each subroutine required, NOT
+into the module header, in order to keep track of module variable modification.
+
+
+### Line numbers
+
+Avoid where possible. Line numbers are only to be used for format statements,
+and `end=` or `err=` in file handling. Do not use `continue`.
+
+
+### General features
+
+Avoid multiple statement lines unless trivially simple.
+
+
+## Modules
+
+
+### General
+
+All code is to be contained within modules.
+
+
+### Usage
+
+Each module described in the specification is to be one source file. All
+routines given in the specification for this module are to be in this file.
+
+
+### Auxiliary modules
+
+For clarity, and to keep the module from being excessively long, the module may
+also use auxiliary modules, which can be in separate files (again with one file
+per module). Any auxiliary modules must be used in only one specified module,
+i.e. no auxiliary code reuse in other specified modules as the auxiliary
+routines are not specified! The function of the auxiliary module must, of
+course, be documented in the Implementation Guide.
+
+The naming convention of any auxiliary modules is to be inherited from the
+specified module to make it clear where the auxiliary routines are to be used,
+e.g. if the specified module is called `me.F90` with routines called `me_sub`,
+etc. then an auxiliary module would be `me_extra.F90` with routines called
+`me_extra_sub`, etc.
+
+
+## Style and Layout
+
+
+### Use `ModuleTemplate.F90`
+
+The `ModuleTemplate.F90` file gives the recommended layout for a module and for
+a subprogram. In particular, note the layout and contents of the comment header,
+and the order of declarations. Everything is to be private unless explicitly
+specified as public! This also makes it simple to check the code against the
+specification.
+
+
+### Layout
+
+All code is to use indentation as generated by GNU-Emacs f90-mode. Use any
+editor you like to create the code, but please ensure that whatever goes into
+the Bitbucket repository has been passed through Emacs f90-mode indentation at the
+finished stage. To assist in Emacs usage, the `ModuleTemplate.F90` file contains
+Emacs directives to set f90-mode, etc.
+
+
+### Case
+
+Use lower case for keywords and use capitalisation as given in the specification
+for all routines and variables mentioned therein. For private routines and variables
+use lower case in general. However, you may use upper case to differentiate
+words within a name, e.g. `userInputData` - as long as you are consistent
+throughout the module.
```

Comment:
Think we want to specify to use `lower_snake` case generally.

Reply:
It is optional. Some prefer it, others done. Hence if you WANT to use uppercase within a name then you can, but 'lower case in general' is CASTEP style.